src/xma/src/xmaapi/xmadecoder.cpp

```diff
@@ -246,37 +197,56 @@ xma_dec_session_create(XmaDecoderProperties *dec_props)
     priv1->kernel_execbos.reserve(num_execbo);
     priv1->num_execbo_allocated = num_execbo;
     if (xma_core::create_session_execbo(priv1, num_execbo, XMA_DECODER_MOD) != XMA_SUCCESS) {
-        //Release singleton lock
-        g_xma_singleton->locked = false;
         free(dec_session->base.plugin_data);
         free(dec_session);
         delete priv1;
         return nullptr;
     }
 
-    if (dec_session->decoder_plugin->init(dec_session)) {
-        xma_logmsg(XMA_ERROR_LOG, XMA_DECODER_MOD,
-                   "Initalization of plugin failed\n");
-        //Release singleton lock
-        g_xma_singleton->locked = false;
-        free(dec_session->base.plugin_data);
-        free(dec_session);
-        delete priv1;
-        return nullptr;
+    //Obtain lock only for a) singleton changes & b) kernel_info changes
+    std::unique_lock<std::mutex> guard1(g_xma_singleton->m_mutex);
+    //Singleton lock acquired
+
+    if (!kernel_info->soft_kernel && !kernel_info->in_use && !kernel_info->context_opened) {
+        if (xclOpenContext(dev_handle, dev_tmp1.uuid, kernel_info->cu_index_ert, true) != 0) {
+            xma_logmsg(XMA_ERROR_LOG, XMA_DECODER_MOD, "Failed to open context to CU %s for this session\n", kernel_info->name);
+            free(dec_session->base.plugin_data);
+            free(dec_session);
+            delete priv1;
+            return nullptr;
+        }
     }
+    dec_session->base.session_id = g_xma_singleton->num_of_sessions + 1;
+    xma_logmsg(XMA_INFO_LOG, XMA_DECODER_MOD,
+                "XMA session channel_id: %d; session_id: %d\n", dec_session->base.channel_id, dec_session->base.session_id);
+
     if (kernel_info->in_use) {
         kernel_info->is_shared = true;
+        xma_logmsg(XMA_DEBUG_LOG, XMA_DECODER_MOD,
+                   "XMA session sharing CU: %s\n", hwcfg->devices[hwcfg_dev_index].kernels[cu_index].name);
     } else {
         kernel_info->in_use = true;
+        xma_logmsg(XMA_DEBUG_LOG, XMA_DECODER_MOD,
+                   "XMA session with CU: %s\n", hwcfg->devices[hwcfg_dev_index].kernels[cu_index].name);
     }
     kernel_info->num_sessions++;
     g_xma_singleton->num_decoders++;
     g_xma_singleton->num_of_sessions = dec_session->base.session_id;
 
-    g_xma_singleton->all_sessions.emplace(g_xma_singleton->num_of_sessions, dec_session->base);
+    g_xma_singleton->all_sessions_vec.push_back(dec_session->base);
 
     //Release singleton lock
-    g_xma_singleton->locked = false;
+    guard1.unlock();

```

Comment:
Something to think about, instead of declaring an unlock, you could scope the mutext lock as follows:
```
...
{
   std::unique_lock<std::mutex> guard1(g_xma_singleton->m_mutex);

    if (!kernel_info->soft_kernel && !kernel_info->in_use && !kernel_info->context_opened) {
        if (xclOpenContext(dev_handle, dev_tmp1.uuid, kernel_info->cu_index_ert, true) != 0) {
            xma_logmsg(XMA_ERROR_LOG, XMA_DECODER_MOD, "Failed to open context to CU %s for this session\n", kernel_info->name);
            free(dec_session->base.plugin_data);
            free(dec_session);
            delete priv1;
            return nullptr;
        }
    }
    dec_session->base.session_id = g_xma_singleton->num_of_sessions + 1;
    xma_logmsg(XMA_INFO_LOG, XMA_DECODER_MOD,
                "XMA session channel_id: %d; session_id: %d\n", dec_session->base.channel_id, dec_session->base.session_id);

    if (kernel_info->in_use) {
        kernel_info->is_shared = true;
        xma_logmsg(XMA_DEBUG_LOG, XMA_DECODER_MOD,
                   "XMA session sharing CU: %s\n", hwcfg->devices[hwcfg_dev_index].kernels[cu_index].name);
    } else {
        kernel_info->in_use = true;
        xma_logmsg(XMA_DEBUG_LOG, XMA_DECODER_MOD,
                   "XMA session with CU: %s\n", hwcfg->devices[hwcfg_dev_index].kernels[cu_index].name);
    }
    kernel_info->num_sessions++;
    g_xma_singleton->num_decoders++;
    g_xma_singleton->num_of_sessions = dec_session->base.session_id;

    g_xma_singleton->all_sessions.emplace(g_xma_singleton->num_of_sessions, dec_session->base);
    g_xma_singleton->all_sessions_vec.push_back(dec_session->base);
}
...
```

Doing so eliminates the need to call unlock.   Just an idea.